### ISSUE TITLE
1. polymorphism? 2. action bubbling?

### ISSUE THREAD

**Issue by hex13** [rackt/react-redux#160]
Hello, 
I'm using Redux(+React) to make operate on hierarchical structures and I have two questions:
1. how to implement polymorphism in Redux?
2. how to make "action bubbling"

My questions expressed more precisely:
1. I'm making structural code editor in React, using Redux. By structural editor I mean that I have tree structure with nodes in it and my app is supposed to operate on AST rather than plain text. 

These nodes can have various types (because it's editor for editing CSS files, these types are for example: "rule", "declaration", "property", "value", but it's more important that they do have some type at all). So there are objects in JS which have property `type` and its `type` determines its... type ;)

So assume that I have node like this:
`{type: 'property', id: 4, rule: '.content > ul > li', children: [...blah blah......]}`

And I would like to be able write something like that:
`store.dispatch({type: 'click', nodeId: 4})`
for click. But I would expect that clicking on node of type `rule` would have other effect than clicking on node of type `property`. Polymorphism (but I would like to have ONE react component `<Node />` to prevent code duplication).

I know I can achieve that either by switch/case or making my own sub-functions (I mean to delegate action to functions), but I don't want reinvent the the wheel. I think that support of polymorphism would be useful to many applications, not only mine. 

So - how to achieve polymorphism in Redux? 

 
2., Second question about action bubbling. Because I got tree structure for example such like this one:
```
{
    type: 'rule',
    rule: 'body div',
    id: 1,
    children: [
         {
              type: 'declaration',
              id: 2,
              children: [
                    {type: 'property', name: 'display', id: 3},
                    {type: 'value', value: 'block', id: 4},
              ]
         } 
    ]
}
```
And I have appropriate React components that render that structure (I have dumb function-based stateless React 0.14 components, so I don't think components' code is very important right now). Anyway I got `<Node/>` component that represents single AST node.

I would like that user could click on for example on node at id=3 ( ` {type: 'property', name: 'display', id: 3},`) and action would automatically "bubble" to node's parents (in this case to declaration node at id=2  and to rule node at id=1).

As far I know there is no "action bubbling" in Redux. Maybe it should be? Or maybe what I would like to achieve could be made by other means? (by "action bubbling" I mean something similar to "event bubbling" in JavaScript events in browser, only I would like `redux actions` to bubble and I do NOT want DOM events to bubble (this is completely other thing). I would like to redux action to bubble. When I wrote "bubbling" I ment overall design pattern rather than implementation.

And these were my two questions :)

**Comment by gaearon** [comment]
Hi, please ask this on StackOverflow instead. It is likely you'll get a much better answer there because this is really a usage question, not API question or issue.

**Comment by hex13** [comment]
Actually I thought it would be nice to have support for these two features (polymorphism and action bubbling) in Redux or React-Redux itself (if it's possible).

But okay, I wrote question on Action Overflow:
http://stackoverflow.com/questions/33169415/polymorphism-and-action-bubbling-in-redux

**Comment by gaearon** [comment]
To be honest I don't understand neither what you mean by polymorphism nor what you mean by bubbling. :-(

Maybe if you can provide a specific runnable example and show where Redux doesn't currently satisfies it, I can help you more.

**Comment by hex13** [comment]
TL;DR: 
- I solved my problem (using higher-order components and higher-order functions in general). I done this on the React level (this turned out to be not Redux specific)
- I'm gonna probably make mini-library from my solution and put this on github
- But I'll write below what I meant previously, just to be clear.

`To be honest I don't understand neither what you mean by polymorphism`
I'm just trying to apply standard OOP patterns to functional and data oriented programming. 

So. I have some hierarchical tree like data structure (like in my first post but instead of CSS AST imagine maybe that I build something more normal like Forum Engine with subforums, threads and posts). And I render this tree in React:
```javascript
const Node = ({node})  => (
   <div>

      //  render some arbitrary view inside the node based on node.type
      //   every node of the tree should look and behave according to its "type" property. 
       //  
      // ......

      // visit children
      <ul>
      {node.children.map(
          item => <li key={item.id}><Node node={item}/></li>
       )}
       </ul>
   </div>
)
```

Nothing weird here, just implementing tree structure and visiting children, and recursive rendering subtrees.

The problem was with this part
```javascript
   <div>
      // .... (render some arbitrary view inside the node based on node.type )
      // ......
```
I didn't want end up with big switch like this:
```
switch (node.type) {
   case 'subforum': <a href="{node.link}">{ node.name }  { node.newestThread.name }</a>; break;
   case 'thread': <a href="{node.link}">{node.name} </a>; break;
   case 'post': <div><h3>{node.author} - {node.date}</h3>{ node.content }</div>
}
```
I never really wanted to use `switch/case`, but just to illustrate. I just knew that I had to pack all this cases in some separate functions. And this is why I used word `polymorphism`.  Finally I made some simple higher level components for resolve my problems :)

It turned out also that second thing - i.e. what I wanted achieve by "bubbling" can be achieved in passing handlers to components (like it's in Redux-React itself but previously I just didn't like the idea. Maybe I was just wrong in this case)
